src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs

```diff
@@ -418,10 +450,17 @@ private void RegisterFor(int ev)
             }
 
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            private async void CleanupSocketEnd()
+            private void CleanupSocketEnd()
             {
+                var hasCanceledConnectionClosedToken = false;
```

Comment:
Do we need `_waitForConnectionClosedTcs` the `if (!HasFlag(SocketFlags.BothClosed))` will return for the first caller, and succeed for the second.

Reply:
I cleaned this up just to check if `HasFlag(SocketFlags.BothClosed)` was already true at the top of `CleanupSocketEnd()` and return early in that case. Otherwise I would get an error like the following when running the sample app, since `CleanupSocketEnd()` can apparently be called after the socket has already been fully closed:

```
System.ObjectDisposedException: The CancellationTokenSource has been disposed.
   at System.Threading.CancellationTokenSource.ThrowObjectDisposedException()
   at System.Threading.CancellationTokenSource.Cancel()
   at RedHat.AspNetCore.Server.Kestrel.Transport.Linux.TransportThread.TSocket.CancelConnectionClosedToken() in /home/shalter/halter73/kestrel-linux-transport/src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs:line 487
Unhandled exception. System.ObjectDisposedException: The CancellationTokenSource has been disposed.
   at System.Threading.CancellationTokenSource.ThrowObjectDisposedException()
   at System.Threading.CancellationTokenSource.Cancel()
   at RedHat.AspNetCore.Server.Kestrel.Transport.Linux.TransportThread.TSocket.CancelConnectionClosedToken() in /home/shalter/halter73/kestrel-linux-transport/src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs:line 487
   at RedHat.AspNetCore.Server.Kestrel.Transport.Linux.TransportThread.TSocket.<>c.<CleanupSocketEnd>b__63_0(Object state) in /home/shalter/halter73/kestrel-linux-transport/src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs:line 473
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

Reply:
I don't understand why this is throwing for you., this is how it is supposed to work:

For each socket, `CleanupSocketEnd` is called twice. The first time it counts and no-ops. The second time it queues, `CancelConnectionClosedToken` to the `ThreadPool`.
`CancelConnectionClosedToken` shouldn't throw (like in this stacktrace), because the token is disposed in `DisposeAsync` after awaiting `_waitForConnectionClosedTcs` (which is set from `CancelConnectionClosedToken`).

Reply:
Try changing https://github.com/redhat-developer/kestrel-linux-transport/pull/80#pullrequestreview-276548042, this is currently causing an extra call to `CleanupSocketEnd`.